drivers/watchdog/wdt_gecko.c

```diff
@@ -0,0 +1,285 @@
+struct wdt_gecko_cfg {
+	WDOG_TypeDef *base;
+	void (*irq_cfg_func)(void);
+};
+
+struct wdt_gecko_data {
+	wdt_callback_t callback;
+	WDOG_Init_TypeDef wdog_config;
+	bool timeout_valid;
+};
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_DATA(dev) \
+	((struct wdt_gecko_data *)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((struct wdt_gecko_cfg *)(dev)->config->config_info)
+
+static int wdt_gecko_convert_timeout(u32_t timeout)
+{
+	int idx = 0;
+
+	/* When using ULFRCO (default), 1 cycle is 1 ms +/- 12% */
+	while (idx < ARRAY_SIZE(timeout_in_cycles_tbl)) {
+		if (timeout > timeout_in_cycles_tbl[idx]) {
+			idx++;
+			continue;
+		}
+
+		/* Found the (rounded up?) value of cycles for
+		 *  supplied timeout
+		 */
+		break;
+	}
+
+	return idx;
+}
+
+static int wdt_gecko_convert_window(u32_t window, u32_t period)
+{
+	int idx = 0;
+	u32_t incr_val, comp_val;
+
+	incr_val = period / 8;
+	comp_val = 0; /* Initially 0, disable */
+
+	/* Valid window settings range from 12.5% of the calculated
+	 * timeout period up to 87.5% (= 7 * 12.5%)
+	 */
+	while (idx < 7) {
+		if (window > comp_val) {
+			comp_val += incr_val;
+			idx++;
+			continue;
+		}
+
+		break;
+	}
+
+	return idx;
+}
+
+static int wdt_gecko_setup(struct device *dev, u8_t options)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	if (!data->timeout_valid) {
+		LOG_ERR("No valid timeouts installed");
+		return -EINVAL;
+	}
+
+	data->wdog_config.em2Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+	data->wdog_config.em3Run =
+		(options & WDT_OPT_PAUSE_IN_SLEEP) == 0U;
+
+	data->wdog_config.debugRun =
+		(options & WDT_OPT_PAUSE_HALTED_BY_DBG) == 0U;
+
+	if (data->callback != NULL) {
+		/* Interrupt mode for window */
+		/* Enable timeout interrupt */
+		WDOGn_IntEnable(wdog, WDOG_IEN_TOUT);
+	} else {
+		/* Disable timeout interrupt */
+		WDOGn_IntDisable(wdog, WDOG_IEN_TOUT);
+	}
+
+	/* Watchdog is started after initialization */
+	WDOGn_Init(wdog, &data->wdog_config);
+	LOG_DBG("Setup the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_disable(struct device *dev)
+{
+	const struct wdt_gecko_cfg *config = DEV_CFG(dev);
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_TypeDef *wdog = config->base;
+
+	WDOGn_Enable(wdog, false);
+	data->timeout_valid = false;
+	LOG_DBG("Disabled the watchdog");
+
+	return 0;
+}
+
+static int wdt_gecko_install_timeout(struct device *dev,
+				     const struct wdt_timeout_cfg *cfg)
+{
+	struct wdt_gecko_data *data = DEV_DATA(dev);
+	WDOG_Init_TypeDef init_defaults = WDOG_INIT_DEFAULT;
+
+	if (data->timeout_valid) {
+		LOG_ERR("No more timeouts can be installed");
+		return -ENOMEM;
+	}
+
+	if ((cfg->window.max < timeout_in_cycles_tbl[0]) ||
+		(cfg->window.max > timeout_in_cycles_tbl[
+			ARRAY_SIZE(timeout_in_cycles_tbl) - 1])) {
+		LOG_ERR("Upper limit timeout out of range");
+		return -EINVAL;
+	}
```

Comment:
The actual timeout period set by the watchdog will be widely inaccurate compared to the required value. E.g. given 131074 timeout value provided via the API call the watchdog will expire after 262145. As such there is no reason to reject value 262146. I propose to remove this check entirely. If we are concerned about the accuracy of the timeout setting we need to extend the watchdog API. E.g. provide an extra function that will allow application to read out the timeout value set by the watchdog.